src/hierarchies/mfa_hierarchy.cc

```diff
@@ -0,0 +1,286 @@
+  data_sum = Eigen::VectorXd::Zero(dim);
+}
+
+void MFAHierarchy::set_state_from_proto(
+    const google::protobuf::Message& state_) {
+  auto& statecast = downcast_state(state_);
+  state.mu = bayesmix::to_eigen(statecast.mfa_state().mu());
+  state.psi = bayesmix::to_eigen(statecast.mfa_state().psi());
+  state.eta = bayesmix::to_eigen(statecast.mfa_state().eta());
+  state.lambda = bayesmix::to_eigen(statecast.mfa_state().lambda());
+  state.psi_inverse = state.psi.cwiseInverse().asDiagonal();
+  set_card(statecast.cardinality());
+}
+
+std::shared_ptr<bayesmix::AlgorithmState::ClusterState>
+MFAHierarchy::get_state_proto() const {
+  bayesmix::MFAState state_;
+  bayesmix::to_proto(state.mu, state_.mutable_mu());
+  bayesmix::to_proto(state.psi, state_.mutable_psi());
+  bayesmix::to_proto(state.eta, state_.mutable_eta());
+  bayesmix::to_proto(state.lambda, state_.mutable_lambda());
+  bayesmix::to_proto(state.psi_inverse, state_.mutable_psi_inverse());
+
+  auto out = std::make_shared<bayesmix::AlgorithmState::ClusterState>();
+  out->mutable_mfa_state()->CopyFrom(state_);
+  return out;
+}
+
+void MFAHierarchy::set_hypers_from_proto(
+    const google::protobuf::Message& hypers_) {
+  auto& hyperscast = downcast_hypers(hypers_).mfa_state();
+  hypers->mutilde = bayesmix::to_eigen(hyperscast.mutilde());
+  hypers->alpha0 = hyperscast.alpha0();
+  hypers->beta = bayesmix::to_eigen(hyperscast.beta());
+  hypers->phi = hyperscast.phi();
+  hypers->q = hyperscast.q();
+}
+
+std::shared_ptr<bayesmix::AlgorithmState::HierarchyHypers>
+MFAHierarchy::get_hypers_proto() const {
+  bayesmix::MFAPriorDistribution hypers_;
+  bayesmix::to_proto(hypers->mutilde, hypers_.mutable_mutilde());
+  bayesmix::to_proto(hypers->beta, hypers_.mutable_beta());
+  hypers_.set_alpha0(hypers->alpha0);
+  hypers_.set_phi(hypers->phi);
+  hypers_.set_q(hypers->q);
+
+  auto out = std::make_shared<bayesmix::AlgorithmState::HierarchyHypers>();
+  out->mutable_mfa_state()->CopyFrom(hypers_);
+  return out;
+}
+
+void MFAHierarchy::sample_full_cond(bool update_params) {
+  if (this->card == 0) {
+    // No posterior update possible
+    sample_prior();
+  } else {
+    sample_eta();
+    sample_mu();
+    sample_psi();
+    sample_lambda();
+  }
+}
+
+void MFAHierarchy::sample_eta() {
+  auto& rng = bayesmix::Rng::Instance().get();
+  auto sigma_eta_inv_llt =
+      (Eigen::MatrixXd::Identity(hypers->q, hypers->q) +
+       state.lambda.transpose() * state.psi_inverse * state.lambda)
+          .llt();
+  Eigen::MatrixXd sigma_eta(sigma_eta_inv_llt.solve(
+      Eigen::MatrixXd::Identity(hypers->q, hypers->q)));
+
+  if (state.eta.rows() != card) {
+    state.eta.resize(card, state.eta.cols());
+    state.eta = Eigen::MatrixXd::Zero(card, state.eta.cols());
+  }
+  Eigen::MatrixXd temp_product(sigma_eta * (state.lambda.transpose()) *
+                               state.psi_inverse);
+  auto iterator = cluster_data_idx.begin();
```

Comment:
Using an iterator on the `std::set` we only access data points we need inside the cluster.